Vision_feats/generate_feats.py

```diff
@@ -0,0 +1,308 @@
+            img_names = [
+                i.strip('\n') for i in open(img_name_file, 'r').readlines()
+            ]
+            self.dataset = [
+                os.path.join(args.raw_image_dir, i) for i in img_names
+            ]
+        else:
+            pass
+        self.num_grayscale = 0
+
+    def load_imagepaths_with_labels(self) -> List[str]:
+        """Fetches all images in the dataset.
+        """
+        img_paths = []
+        img_paths = glob.glob(self.data_dir + '/*.jpg')
+        return img_paths
+
+    def __getitem__(self, index: int) -> Tuple[torch.Tensor, int]:
+        """Fetches the item at a given index.
+        """
+        img = None
+        path = self.dataset[index]
+        with open(path, 'rb') as f:
+            img = Image.open(f)
+            if img.mode != 'RGB':
+                self.num_grayscale += 1
+                img = img.convert('RGB')
+            img = self.transform(img)
+        return img
+
+    def __len__(self) -> int:
+        """Returns the number of items in the dataset.
+        """
+        l = len(self.dataset)
+        return l
+
+
+class MyResNet50(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        model = resnet50(pretrained=True)
+        self.conv_layers = nn.Sequential(*list(model.children())[:-1])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        model_output = self.conv_layers(x)
+        return model_output
+
+
+class MyBeit(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        #self.feat_ex = BeitFeatureExtractor.from_pretrained("microsoft/beit-large-patch16-224")
+        self.model = BeitModel.from_pretrained(
+            models_file[args.img_encode_model]
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        #x = self.feat_ex(x)
+        x = {'pixel_values': x}
+        model_output = self.model(**x)
+        model_output = model_output.last_hidden_state
+        return model_output
+
+
+class MyClip(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        #self.feat_ex = BeitFeatureExtractor.from_pretrained("microsoft/beit-large-patch16-224")
+        full_model, _ = clip.load(models_file[args.img_encode_model]) 
+        #self.model = full_model.visual
+        self.model = full_model
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        #x = self.feat_ex(x)
+        model_output = self.model.encode_image(x).float()
+        return model_output
+
+
+class VisionEncode(nn.Module):
```

Comment:
I also feel like this might be better with subclasses.  OR minimally, making downsampler methods an argument to this initializer, instead of all of the if/else based on strings in one big downsample method

Reply:
See above ^ :)